parser/src/ast/helper.rs

```diff
@@ -118,140 +98,234 @@ impl NumberCompare for f64 {
+///This type if for comparing Simple<Factor> to Simple<MathExpr> and returning
+/// Simples, this makes sure that only correct simples can be constructed
+pub trait SimpleCompareFactor {
+    ///gets the math_exprs to compare
+    fn to_expr(&self) -> (&Factor, &MathExpr);
+    ///pows them and produces a Simple
+    fn pow_wrapped(self) -> Simple<Factor>;
+}
+impl SimpleCompareFactor for (Simple<Factor>, Simple<MathExpr>) {
+    fn to_expr(&self) -> (&Factor, &MathExpr) {
+        (&self.0, &self.1)
     }
 
+    fn pow_wrapped(self) -> Simple<Factor> {
+        Simple(Factor::Power {
+            base: self.0.boxed_inner(),
+            exponent: self.1.boxed_inner(),
+        })
+    }
+}
+impl SimpleCompareEquivalent for (Simple<Factor>, Simple<MathExpr>) {
     fn equivalent(&self, cont: &MathContext) -> bool {
-        self.1.equivalent(&self.0, cont)
+        self.1.equivalent(
+            &Simple(MathExpr::Term(Term::Factor(self.0 .0.clone()))),
+            cont,
+        )
     }
 }
+///This type if for comparing Simple<MathExpr> to Simple<MathExpr> and
+/// returning Simples, this makes sure that only correct simples can be
+/// constructed
+pub trait SimpleCompareMultipleMathExprs {
+    ///gets the math_exprs to compare
+    fn to_expr(&self) -> (&MathExpr, &MathExpr);
+    /// takes the root as simple
+    fn root(self) -> Simple<Factor>;
+}
+impl SimpleCompareMultipleMathExprs for (Simple<MathExpr>, Simple<MathExpr>) {
+    fn to_expr(&self) -> (&MathExpr, &MathExpr) {
+        (&self.0, &self.1)
+    }
 
+    fn root(self) -> Simple<Factor> {
+        Simple(Factor::Root {
+            degree: Some(self.0.expr().boxed()),
+            radicand: self.1.expr().boxed(),
+        })
+    }
+}
 ///Simple is a wrapper struct only allowed to be constructed when the contained
 /// MathExpr is in the simplest form
 #[derive(Clone, Debug, PartialEq)]
-pub struct Simple(MathExpr);
-impl Simple {
+pub struct Simple<T: Simplify>(pub(crate) T);
+impl<T: Simplify> Simple<T> {
+    ///returns the inner T in a box
+    pub fn boxed_inner(self) -> Box<T> {
+        Box::new(self.0)
+    }
     ///Constructs a Simple from a MathExpr
     pub fn new(
-        math_expr: MathExpr,
+        math_expr: T,
         cont: &MathContext,
-    ) -> Result<Simple, EvalError> {
+    ) -> Result<Simple<T>, EvalError> {
         math_expr.simple(cont)
     }
     /// Construct a Simple without actually checking
     /// that it's simplified.
-    pub const fn new_unchecked(expr: MathExpr) -> Self {
+    pub const fn new_unchecked(expr: T) -> Self {
         Simple(expr)
     }
-    ///Constructs a Ast::Expression from the contained MathExpr
-    pub fn expression(self) -> Ast {
-        Ast::Expression(self.0)
-    }
     ///gets a ref to inner item
-    pub const fn math_expr(&self) -> &MathExpr {
+    pub const fn inner(&self) -> &T {
         &self.0
     }
     ///gets a ref to inner item
-    pub fn expr(self) -> MathExpr {
+    pub fn expr(self) -> T {
```

Comment:
Maybe rename this since it's not always a MathExpr anymore? Maybe inner_move? Or rename inner to inner_ref and call this inner